#                          Assignment - Knn 

# Loading Data

In [1]:
import pandas as pd
import numpy as np

In [2]:
house_data=pd.read_csv("C:/Users/DELL/Downloads/House_data.csv")
house_data

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0    13300000  7420         4          2        3      yes        no       no   
1    12250000  8960         4          4        4      yes        no       no   
2    12250000  9960         3          2        2      yes        no      yes   
3    12215000  7500         4          2        2      yes        no      yes   
4    11410000  7420         4          1        2      yes       yes      yes   
..        ...   ...       ...        ...      ...      ...       ...      ...   
540   1820000  3000         2          1        1      yes        no      yes   
541   1767150  2400         3          1        1       no        no       no   
542   1750000  3620         2          1        1      yes        no       no   
543   1750000  2910         3          1        1       no        no       no   
544   1750000  3850         3          1        2      yes        no       no   

    hotwaterheating airconditioning  parking prefarea furnishingstatus  
0                no             yes        2      yes        furnished  
1                no             yes        3       no        furnished  
2                no              no        2      yes   semi-furnished  
3                no             yes        3      yes        furnished  
4                no             yes        2       no        furnished  
..              ...             ...      ...      ...              ...  
540              no              no        2       no      unfurnished  
541              no              no        0       no   semi-furnished  
542              no              no        0       no      unfurnished  
543              no              no        0       no        furnished  
544              no              no        0       no      unfurnished  

[545 rows x 13 columns]

# Data Prepration

In [3]:
house_data.shape

(545, 13)

In [4]:
house_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


In [5]:
house_data.isnull().sum()

price               0
area                0
bedrooms            0
bathrooms           0
stories             0
mainroad            0
guestroom           0
basement            0
hotwaterheating     0
airconditioning     0
parking             0
prefarea            0
furnishingstatus    0
dtype: int64

In [6]:
house_data.columns

Index(['price', 'area', 'bedrooms', 'bathrooms', 'stories', 'mainroad',
       'guestroom', 'basement', 'hotwaterheating', 'airconditioning',
       'parking', 'prefarea', 'furnishingstatus'],
      dtype='object')

In [7]:
var=['mainroad','guestroom', 'basement', 'hotwaterheating', 'airconditioning','prefarea']
house_data[var]=house_data[var].replace(to_replace={'yes':1,'no':0})
house_data.head()

price  area  bedrooms  bathrooms  stories  mainroad  guestroom  \
0  13300000  7420         4          2        3         1          0   
1  12250000  8960         4          4        4         1          0   
2  12250000  9960         3          2        2         1          0   
3  12215000  7500         4          2        2         1          0   
4  11410000  7420         4          1        2         1          1   

   basement  hotwaterheating  airconditioning  parking  prefarea  \
0         0                0                1        2         1   
1         0                0                1        3         0   
2         1                0                0        2         1   
3         1                0                1        3         1   
4         1                0                1        2         0   

  furnishingstatus  
0        furnished  
1        furnished  
2   semi-furnished  
3        furnished  
4        furnished

In [8]:
house_data['furnishingstatus']=house_data['furnishingstatus'].replace(to_replace={'furnished':1,'semi-furnished':2,'unfurnished':0})
house_data.head()

price  area  bedrooms  bathrooms  stories  mainroad  guestroom  \
0  13300000  7420         4          2        3         1          0   
1  12250000  8960         4          4        4         1          0   
2  12250000  9960         3          2        2         1          0   
3  12215000  7500         4          2        2         1          0   
4  11410000  7420         4          1        2         1          1   

   basement  hotwaterheating  airconditioning  parking  prefarea  \
0         0                0                1        2         1   
1         0                0                1        3         0   
2         1                0                0        2         1   
3         1                0                1        3         1   
4         1                0                1        2         0   

   furnishingstatus  
0                 1  
1                 1  
2                 2  
3                 1  
4                 1

In [9]:
from sklearn.model_selection import train_test_split

In [10]:
train,test=train_test_split(house_data,train_size=0.7,test_size=0.3,random_state=100)
train.head()

price  area  bedrooms  bathrooms  stories  mainroad  guestroom  \
359  3710000  3600         3          1        1         1          0   
19   8855000  6420         3          2        2         1          0   
159  5460000  3150         3          2        1         1          1   
35   8080940  7000         3          2        4         1          0   
28   8400000  7950         5          2        2         1          0   

     basement  hotwaterheating  airconditioning  parking  prefarea  \
359         0                0                0        1         0   
19          0                0                1        1         1   
159         1                0                1        0         0   
35          0                0                1        2         0   
28          1                1                0        2         0   

     furnishingstatus  
359                 0  
19                  2  
159                 1  
35                  1  
28                  0

In [11]:
train.shape

(381, 13)

In [12]:
test.shape

(164, 13)

In [13]:
x_train = train.drop('price', axis=1)
y_train = train['price']

x_test = test.drop('price', axis=1)
y_test = test['price']

In [14]:
print(x_train.shape)
print(y_train.shape)
print(x_test.shape)
print(y_test.shape)

(381, 12)
(381,)
(164, 12)
(164,)


In [15]:
x_train 

area  bedrooms  bathrooms  stories  mainroad  guestroom  basement  \
359  3600         3          1        1         1          0         0   
19   6420         3          2        2         1          0         0   
159  3150         3          2        1         1          1         1   
35   7000         3          2        4         1          0         0   
28   7950         5          2        2         1          0         1   
..    ...       ...        ...      ...       ...        ...       ...   
526  3180         2          1        1         1          0         0   
53   5150         3          2        4         1          0         0   
350  3420         2          1        2         1          0         0   
79   6000         3          2        3         1          1         0   
520  7700         2          1        1         1          0         0   

     hotwaterheating  airconditioning  parking  prefarea  furnishingstatus  
359                0                0        1         0                 0  
19                 0                1        1         1                 2  
159                0                1        0         0                 1  
35                 0                1        2         0                 1  
28                 1                0        2         0                 0  
..               ...              ...      ...       ...               ...  
526                0                0        0         0                 0  
53                 0                1        2         0                 2  
350                1                0        1         0                 2  
79                 0                1        0         0                 1  
520                0                0        0         0                 0  

[381 rows x 12 columns]

In [16]:
y_train

359    3710000
19     8855000
159    5460000
35     8080940
28     8400000
        ...   
526    2310000
53     7350000
350    3780000
79     6650000
520    2450000
Name: price, Length: 381, dtype: int64

In [17]:
x_test

area  bedrooms  bathrooms  stories  mainroad  guestroom  basement  \
265   2880         3          1        2         1          0         0   
54    6000         3          2        2         1          1         0   
171  10269         3          1        1         1          0         0   
244   5320         3          1        2         1          1         1   
268   4950         4          1        2         1          0         0   
..     ...       ...        ...      ...       ...        ...       ...   
413   1950         3          2        2         1          0         1   
191  10700         3          1        2         1          1         1   
310   5985         3          1        1         1          0         1   
46    6000         3          2        4         1          0         0   
50    7440         3          2        4         1          0         0   

     hotwaterheating  airconditioning  parking  prefarea  furnishingstatus  
265                0                0        0         1                 2  
54                 0                1        1         0                 2  
171                0                0        1         1                 2  
244                0                0        0         1                 2  
268                0                1        0         0                 2  
..               ...              ...      ...       ...               ...  
413                0                0        0         1                 0  
191                0                0        0         0                 2  
310                0                0        0         0                 2  
46                 0                1        1         0                 1  
50                 0                0        1         1                 0  

[164 rows x 12 columns]

# Knn 

In [18]:
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error

In [19]:
rmse=[] 
for i in range(0,8):
    i = i+1
    knn = KNeighborsRegressor(n_neighbors=i)

    knn.fit(x_train,y_train)  
    pred=knn.predict(x_test) 

    rms =np.sqrt(mean_squared_error(y_test,pred)) 
    rmse.append(rms)
    print("root_mean_square_error for k=",i,'is:',rms)
    
     

root_mean_square_error for k= 1 is: 2168251.9318511444
root_mean_square_error for k= 2 is: 1829803.6767124406
root_mean_square_error for k= 3 is: 1712776.7906637997
root_mean_square_error for k= 4 is: 1630970.4496991842
root_mean_square_error for k= 5 is: 1607419.70834821
root_mean_square_error for k= 6 is: 1617085.7230100513
root_mean_square_error for k= 7 is: 1623147.9305627828
root_mean_square_error for k= 8 is: 1618040.8398566274


The minimum root mean square value is minimum here for k=4,5,6,7,8
 Hence,we have RMSE is 1607419.70834821 for k=5

In [20]:
from sklearn.metrics import mean_absolute_error

In [21]:
for i in range(0,8):
    i = i+1
    knn = KNeighborsRegressor(n_neighbors=i)

    knn.fit(x_train,y_train)  
    y_pred=knn.predict(x_test) 

    MAE=mean_absolute_error(y_test,y_pred)
    print("mean absolute error for k=",i,'is:',MAE)
    

mean absolute error for k= 1 is: 1566703.719512195
mean absolute error for k= 2 is: 1349881.280487805
mean absolute error for k= 3 is: 1270550.9349593495
mean absolute error for k= 4 is: 1189329.131097561
mean absolute error for k= 5 is: 1175313.487804878
mean absolute error for k= 6 is: 1170113.8821138213
mean absolute error for k= 7 is: 1201515.731707317
mean absolute error for k= 8 is: 1198667.7286585367


In [26]:
from sklearn.metrics import r2_score

for i in range(0,8):
    i = i+1
    knn = KNeighborsRegressor(n_neighbors=i)

    knn.fit(x_train,y_train)  
    y_pred=knn.predict(x_test) 

    r2=r2_score(y_test,y_pred)
    print("R square for k=",i,'is:',r2)
    


R square for k= 1 is: -0.19747883664695953
R square for k= 2 is: 0.1471800434885393
R square for k= 3 is: 0.25277756758682135
R square for k= 4 is: 0.3224512511307198
R square for k= 5 is: 0.3418771944161231
R square for k= 6 is: 0.33393832012532454
R square for k= 7 is: 0.3289350322420953
R square for k= 8 is: 0.3331512813242593
